external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,458 @@
+
+    private boolean commit() {
+        return !consumerAutoCommitMode && timer.isExpiredResetOnTrue();    // timer != null for non auto commit mode
+    }
+
+    private ConsumerRecords<K, V> poll() {
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        LOG.debug("Polled [{}] records from Kafka", consumerRecords.count());
+        return consumerRecords;
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());
+
+            for (final ConsumerRecord<K, V> record : records) {
+                if (record.offset() == 0 || consumerAutoCommitMode || record.offset() > acked.get(tp).committedOffset) {      // The first poll includes the last committed offset. This if avoids duplication
+                    final List<Object> tuple = tupleBuilder.buildTuple(record);
+                    final MessageId messageId = new MessageId(record, tuple);
+
+                    kafkaSpoutStreams.emit(collector, messageId);           // emits one tuple per record
+                    LOG.debug("Emitted tuple [{}] for record [{}]", tuple, record);
+                }
+            }
+        }
+    }
+
+    private void commitOffsetsForAckedTuples() {
+        // Find offsets that are ready to be committed for every topic partition
+        final Map<TopicPartition, OffsetAndMetadata> nextCommitOffsets = new HashMap<>();
+        for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+            final OffsetAndMetadata nextCommitOffset = tpOffset.getValue().findNextCommitOffset();
+            if (nextCommitOffset != null) {
+                nextCommitOffsets.put(tpOffset.getKey(), nextCommitOffset);
+            }
+        }
+
+        // Commit offsets that are ready to be committed for every topic partition
+        if (!nextCommitOffsets.isEmpty()) {
+            kafkaConsumer.commitSync(nextCommitOffsets);
+            LOG.debug("Offsets successfully committed to Kafka [{}]", nextCommitOffsets);
+            // Instead of iterating again, it would be possible to commit and update the state for each TopicPartition
+            // in the prior loop, but the multiple network calls should be more expensive than iterating twice over a small loop
+            for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                final OffsetEntry offsetEntry = tpOffset.getValue();
+                offsetEntry.commit(nextCommitOffsets.get(tpOffset.getKey()));
+            }
+        } else {
+            LOG.trace("No offsets to commit. {}", toString());
+        }
+    }
+
+    // ======== Ack =======
+    @Override
+    public void ack(Object messageId) {
+        if (!consumerAutoCommitMode) {  // Only need to keep track of acked tuples if commits are not done automatically
```

Comment:
so if consumerAutoCommitMode is enabled, there is no at least once or at most once processing guarantee? It can be documented if so.